sklearn/tree/_criterion.pyx

```diff
@@ -346,8 +417,77 @@ cdef class ClassificationCriterion(Criterion):
         # Reset to pos=start
         self.reset()
 
+    cdef void init_missing(self, SIZE_t n_missing) nogil:
+        """Separate the total statistics computed at the init, into missing and available statistics
+
+        missing sample indices are stored from samples[end-missing-1:end]
+        """
+        # NOTE init must have been called before
+        self.n_missing = n_missing
+        if n_missing == 0:
+            self.missing_direction = MISSING_DIR_UNDEF
+            return
+
+        cdef SIZE_t end = self.end
+        cdef SIZE_t* samples = self.samples
+        cdef DOUBLE_t* sample_weight = self.sample_weight
+        cdef SIZE_t y_stride = self.y_stride
+        cdef DOUBLE_t* y = self.y
+
+        cdef SIZE_t i
+        cdef SIZE_t p
+        cdef SIZE_t k
+        cdef SIZE_t c
+        cdef DOUBLE_t w = 1.0
+        cdef SIZE_t offset = 0
+
+        # Keep the missing direction as right at the start
+        self.missing_direction = MISSING_DIR_RIGHT
+        self.weighted_n_node_missing = 0.0
+        self.weighted_n_node_available = 0.0
+
+        cdef double* sum_missing = self.sum_missing
+        cdef double* sum_available = self.sum_available
+
+        cdef SIZE_t* n_classes = self.n_classes
+        cdef double* sum_total = self.sum_total
+
+        # SELFNOTE
+        # Why can't we - memset(sum_missing, 0, self.n_outputs * self.sum_stride * sizeof(double))
+        for k in range(self.n_outputs):
+            memset(sum_missing + offset, 0, n_classes[k] * sizeof(double))
+            memset(sum_available + offset, 0, n_classes[k] * sizeof(double))
+            offset += self.sum_stride
+
+        # Compute sum_missing
+        for p in range(end-n_missing, end):
+            i = samples[p]
+
+            # w is originally set to be 1.0, meaning that if no sample weights
+            # are given, the default weight of each sample is 1.0
+            if sample_weight != NULL:
+                w = sample_weight[i]
+
+            # Count weighted class frequency for each target
+            offset = i * y_stride
+
+            for k in range(self.n_outputs):
+                c = <SIZE_t> y[offset + k]
+                sum_missing[k * self.sum_stride + c] += w
+            self.weighted_n_node_missing += w
+
+        # Using the computed sum_missing, separate the sum_available
+        # from the sum_total
+        # {sum_total} --> {sum_available} + {sum_missing}
+        for k in range(self.n_outputs):
+            offset = k * self.sum_stride
+            for c in range(n_classes[k]):
+                i = c + offset
```

Comment:
Ah thanks for the catch. This went unnoticed as I haven't tested on multioutput data :@